stdlib/runtime/concat.gr

```diff
@@ -0,0 +1,64 @@
+/* grainc-flags --no-pervasives */
+module Concat
+
+include "runtime/unsafe/wasmi32"
+include "runtime/unsafe/memory"
+include "runtime/dataStructures"
+from DataStructures use { tagSimpleNumber, allocateArray }
+
+let rec append = (list1, list2) => {
+  match (list1) {
+    [] => list2,
+    [first, ...rest] => [first, ...append(rest, list2)],
+  }
+}
+
+@unsafe
+provide let listConcat = listsArray => {
+  let ptr = WasmI32.fromGrain(listsArray)
+  let length = WasmI32.load(ptr, 4n)
+  let mut result = []
+
+  for (let mut i = 0n; WasmI32.ltU(i, length); i = WasmI32.add(i, 1n)) {
+    let list = listsArray[tagSimpleNumber(i)]
+    result = append(result, list)
+  }
```

Comment:
You'll want to iterate in reverse, since appending is O(length of first list) and the first list is growing. That way total time will be O(k) where k is the sum of the length of all lists.

Reply:
True